### PR TITLE
[processor/transform] introduce aggregate_on_attribute_value function for metrics

### PR DESCRIPTION
**Pull request by odubajDT**
**Link to tracking Issue:** #16224 

**Changes:**

- implemented `aggregate_on_attribute_value` function
- tests
- documentation

**Depends on** https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33669

### CODE REVIEW

File: processor/transformprocessor/README.md [hunk 1]
```diff
@@ -415,6 +416,50 @@ statements:
 
 To aggregate only using a specified set of attributes, you can use `keep_matching_keys`.
 
+### aggregate_on_attribute_value
+
+`aggregate_on_attribute_value(function, attribute, values, newValue)`
+
+The `aggregate_on_attribute_value` function aggreates all datapoints in the metric containing the attribute `attribute` with one of the values present in the `values` parameter using the aggregation function specified in `function` parameter. Additionally it substitutes the values of the specified `attribute` with a new value in `newValue` parameter.
+
+The `aggregate_on_attribute_value` function fetches all datapoints in the metric, which have the attribute `attribute` present with one of the values speficied in `values` and substitutes these values with `newValue`. Afterwards all datapoints are aggregated depending on the attributes.
```
Comment (odubajDT) [inline processor/transformprocessor/README.md:425]:
```suggestion
```

**Comment by odubajDT** [comment]
Hey @TylerHelmuth @evan-bradley would you please have a look?

File: processor/transformprocessor/README.md [hunk 2]
```diff
@@ -415,6 +416,50 @@ statements:
 
 To aggregate only using a specified set of attributes, you can use `keep_matching_keys`.
 
+### aggregate_on_attribute_value
+
+`aggregate_on_attribute_value(function, attribute, values, newValue)`
+
+The `aggregate_on_attribute_value` function aggreates all datapoints in the metric containing the attribute `attribute` with one of the values present in the `values` parameter. `function` is a case-sensitive string that represents the aggregation function.
```
Comment (alecrajeev) [inline processor/transformprocessor/README.md:423]: `aggreates` -> `aggregates`

(4th word in the sentence)

Comment (odubajDT) [inline processor/transformprocessor/README.md:423]:
```suggestion
The `aggregate_on_attribute_value` function aggregates all datapoints in the metric containing the attribute `attribute` with one of the values present in the `values` parameter. `function` is a case-sensitive string that represents the aggregation function.
```

**Comment by odubajDT** [comment]
@TylerHelmuth @evan-bradley would you have time to look into this? Thanks!

**Review by MovieStoreGuy — APPROVED** [review]

**Comment by MovieStoreGuy** [comment]
> Hey @TylerHelmuth @evan-bradley would you please have a look?

Just doubling checking with code owners before it is merged.

File: processor/transformprocessor/README.md [hunk 3]
```diff
@@ -415,6 +416,50 @@ statements:
...
+- min
+- mean
+- median
+- count
+
+**NOTE:** Only the `sum` agregation function is supported for histogram and exponential histogram datatypes.
+
+Examples:
+
+- `aggregate_on_attribute_value(sum, attr1, [val1, val2], new_val) where name == "system.memory.usage`
```
Comment (evan-bradley) [inline processor/transformprocessor/README.md:447]:
```suggestion
- `aggregate_on_attribute_value("sum", "attr1", ["val1", "val2"], "new_val") where name == "system.memory.usage"`
```

File: processor/transformprocessor/README.md [hunk 4]
```diff
@@ -415,6 +416,50 @@ statements:
...
+The `aggregate_on_attribute_value` function can also be used in conjunction with 
+[keep_matching_keys](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs#keep_matching_keys) or
+[delete_matching_keys](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs#delete_matching_keys).
+
+For example, to remove attribute keys matching a regex and aggregate the metrics on the remaining attributes, you can perform the following statement sequence:
+
+```yaml
+statements:
+   - delete_matching_keys(attributes, "(?i).*myRegex.*") where name == "system.memory.usage"
+   - aggregate_on_attribute_value(sum, attr1, [val1, val2], new_val) where name == "system.memory.usage"
```
Comment (evan-bradley) [inline processor/transformprocessor/README.md:458]:
```suggestion
   - aggregate_on_attribute_value("sum", "attr1", ["val1", "val2"], "new_val") where name == "system.memory.usage"
```

File: processor/transformprocessor/README.md [hunk 5]
```diff
@@ -415,6 +416,50 @@ statements:
 
 To aggregate only using a specified set of attributes, you can use `keep_matching_keys`.
 
+### aggregate_on_attribute_value
+
+`aggregate_on_attribute_value(function, attribute, values, newValue)`
+
+The `aggregate_on_attribute_value` function aggregates all datapoints in the metric containing the attribute `attribute` with one of the values present in the `values` parameter. `function` is a case-sensitive string that represents the aggregation function.
+
+Firstly, `attribute` values with one of the values present in `values` are substituted by `newValue` for all datapoints. Afterwards all datapoints are aggregated depending on the attributes.
```
Comment (evan-bradley) [inline processor/transformprocessor/README.md:425]:
```suggestion
The `aggregate_on_attribute_value` function aggregates all datapoints in the metric containing the attribute `attribute` with one of the values present in the `values` into a single datapoint where the attribute has the value `newValue`. `function` is a case-sensitive string that represents the aggregation function.
```

I think we could make this a bit more concise by just explaining how the function uses the parameters to operate on the metric.

File: processor/transformprocessor/internal/metrics/func_agregate_on_attribute_value_metrics_test.go [hunk 6]
```diff
@@ -0,0 +1,581 @@
...
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/aggregateutil"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlmetric"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatatest/pmetrictest"
+)
+
+func Test_aggregateOnAttributeValues(t *testing.T) {
```
Comment (evan-bradley) [inline processor/transformprocessor/internal/metrics/func_agregate_on_attribute_value_metrics_test.go:19]: Could you also add a test with three datapoints where two are aggregated and one is left unaggregated?

**Comment by MovieStoreGuy** [comment]
Hey @evan-bradley, just double checking you're okay with the changes made?

**Review by evan-bradley — APPROVED** [review]

**Comment by evan-bradley** [comment]
This looks good to me, thanks for continuing to push this along @MovieStoreGuy.

@TylerHelmuth could you take a look when you have time?

**Review by TylerHelmuth — APPROVED** [review]

**Pull request merged**